simphony/tools/lattice_tools.py

```diff
@@ -0,0 +1,990 @@
+    p1, p2, p3: array_like
+        primitive vectors
+    tolerance: float
+        tolerance for numerical errors
+
+    Returns
+    -------
+    output : bool
+    '''
+    # there are two types of primitive cells
+    # type 1: (a, 0, 0), (a/2, b/2, 0), (0, 0, c)
+    # type 2: (a/2, -b/2, 0), (a/2, b/2, 0), (0, 0, c)
+
+    lenA, lenB, lenC = map(vector_len, (p1, p2, p3))
+
+    if (numpy.isnan((lenA, lenB, lenC)).any() or
+            numpy.isinf((lenA, lenB, lenC)).any()):
+        raise ValueError("Vectors contain invalid values")
+
+    if numpy.less_equal((lenA, lenB, lenC), tolerance).any():
+        raise ValueError("Edge lengths must be strictly positive")
+
+    dot_products = numpy.abs(map(numpy.dot, (p1, p2, p3), (p2, p3, p1)))
+    right_angles = dot_products <= tolerance
+
+    if right_angles.sum() < 2:
+        return False
+
+    equal_lengths = numpy.abs((lenA-lenB, lenB-lenC, lenC-lenA)) <= tolerance
+
+    if equal_lengths.any() and right_angles.all():
+        return True
+    elif equal_lengths.any():
+        # find the pair of vectors that don't make a right angle
+        # if it is the type 2 cell, they should be equal length
+        if equal_lengths[numpy.where(~right_angles)[0][0]]:
+            return True
+
+    # not type 2 cell, try type 1
+    if not right_angles.all():
+        # find the two vectors that don't make right angles with each other
+        # the projection of one vector is half the another vector
+        not_right_angle = numpy.where(~right_angles)[0][0]
+        dot_product = dot_products[not_right_angle]
+        vec_mapping = {0: (p1, p2), 1: (p2, p3), 2: (p1, p3)}
+        v1, v2 = vec_mapping[not_right_angle]
+        return (numpy.abs(dot_product - numpy.dot(v1, v1)*0.5) < tolerance or
+                numpy.abs(dot_product - numpy.dot(v2, v2)*0.5) < tolerance)
+    else:
+        return False
+
+
+def is_monoclinic_lattice(p1, p2, p3, tolerance=1.e-6):
+    ''' Test if primitive vectors describe a monoclinic lattice
+
+    Also returns True for vectors describing a cubic, hexagonal,
+    tetragonal, orthorhombic or base centered orthorhombic lattice
+
+    Numerical errors are tolerated within ``tolerance``
+
+    Parameters
+    ----------
+    p1, p2, p3: array_like
+        primitive vectors
+    tolerance: float
+        tolerance for numerical errors
+
+    Returns
+    -------
+    output : bool
+    '''
+    lenA, lenB, lenC = map(vector_len, (p1, p2, p3))
+
+    if (numpy.isnan((lenA, lenB, lenC)).any() or
+            numpy.isinf((lenA, lenB, lenC)).any()):
+        raise ValueError("Vectors contain invalid values")
+
+    if numpy.less_equal((lenA, lenB, lenC), tolerance).any():
+        raise ValueError("Edge lengths must be strictly positive")
+
+    cosines = numpy.abs(map(cosine_two_vectors,
+                            (p1, p2, p3), (p2, p3, p1)))
+
+    # base on loose definition: at least 2 angles are 90 degrees
+    return (cosines <= tolerance).sum() >= 2
```

Comment:
I think we should also check here that none of the cosines is equal to 1 or -1. It would guarantee that the primitive cell is 3-dimensional (or has positive volume).